File: lib/entity-resolution.ts
import { institutionTrackingMatches } from "@/lib/institution-tracking-identity";
import rawDecisions from "@/config/entity_resolution_decisions.json";
import { people as catalogPeople } from "@/lib/catalog-data";
import { companyRegistryEntries } from "@/lib/company-registry";
import { userTrackingConfig } from "@/lib/user-tracking";

export type EntityResolutionEntityType = "company" | "person" | "topic";
export type EntityResolutionStatus = "resolved" | "review" | "rejected";
export type EntityResolutionConfidence = "verified" | "high" | "medium" | "low";
export type EntityResolutionSource =
  | "human-decision"
  | "company-registry"
  | "institution-directory"
  | "people-registry"
  | "tracking-taxonomy"
  | "source-context"
  | "explicit-type"
  | "unresolved";

export type EntityResolutionDecision = {
  status: EntityResolutionStatus;
  requestedType: EntityResolutionEntityType;
  entityType: EntityResolutionEntityType;
  canonicalName: string;
  targetId: string;
  aliases: string[];
  confidence: EntityResolutionConfidence;
  note: string;
  reviewedBy: string;
  reviewedAt: string;
};

export type EntityResolutionManifest = {
  schemaVersion: 1;
  generatedAt: string;
  decisions: Record<string, EntityResolutionDecision>;
};

export type EntityResolutionSourceContext = {
  title?: string;
  summary?: string;
  sourceName?: string;
  channel?: string;
  channelLabel?: string;
  eventType?: string;
  url?: string;
};

export type TrackingEntityResolution = {
  status: EntityResolutionStatus;
  requestedType: EntityResolutionEntityType;
  entityType: EntityResolutionEntityType;
  canonicalName: string;
  targetId: string;
  confidence: EntityResolutionConfidence;
  source: EntityResolutionSource;
  reason: string;
  decisionKey: string;
  reclassified: boolean;
};

export type ResolveTrackingEntityInput = {
  requestedType: EntityResolutionEntityType;
  name: string;
  source?: EntityResolutionSourceContext;
  manifest?: EntityResolutionManifest;
};

const ENTITY_TYPES: EntityResolutionEntityType[] = ["company", "person", "topic"];
const STATUSES: EntityResolutionStatus[] = ["resolved", "review", "rejected"];
const CONFIDENCES: EntityResolutionConfidence[] = ["verified", "high", "medium", "low"];

function text(value: unknown, limit = 1_200) {
  return String(value ?? "").normalize("NFKC").replace(/\s+/gu, " ").trim().slice(0, limit);
}

function unique(values: unknown, limit = 40) {
  if (!Array.isArray(values)) return [];
  const result: string[] = [];
  const seen = new Set<string>();
  for (const raw of values) {
    const value = text(raw, 240);
    const key = normalizeEntityResolutionIdentity(value);
    if (!value || !key || seen.has(key)) continue;
    result.push(value);
    seen.add(key);
    if (result.length >= limit) break;
  }
  return result;
}

export function normalizeEntityResolutionIdentity(value: unknown) {
  return text(value, 300)
    .toLocaleLowerCase("zh-CN")
    .replace(/[^a-z0-9\u3400-\u9fff]+/gu, "");
}

function normalizeDecision(value: unknown): EntityResolutionDecision | null {
  if (!value || typeof value !== "object") return null;
  const row = value as Record<string, unknown>;
  const status = STATUSES.includes(row.status as EntityResolutionStatus)
    ? (row.status as EntityResolutionStatus)
    : null;
  const requestedType = ENTITY_TYPES.includes(row.requestedType as EntityResolutionEntityType)
    ? (row.requestedType as EntityResolutionEntityType)
    : "company";
  const entityType = ENTITY_TYPES.includes(row.entityType as EntityResolutionEntityType)
    ? (row.entityType as EntityResolutionEntityType)
    : requestedType;
  const canonicalName = text(row.canonicalName, 160);
  if (!status || !canonicalName) return null;
  return {
    status,
    requestedType,
    entityType,
    canonicalName,
    targetId: text(row.targetId, 240),
    aliases: unique(row.aliases, 30),
    confidence: CONFIDENCES.includes(row.confidence as EntityResolutionConfidence)
      ? (row.confidence as EntityResolutionConfidence)
      : status === "resolved"
        ? "verified"
        : "low",
    note: text(row.note, 600),
    reviewedBy: text(row.reviewedBy, 120),
    reviewedAt: text(row.reviewedAt, 80),
  };
}

export function normalizeEntityResolutionManifest(value: unknown): EntityResolutionManifest {
  const row = value && typeof value === "object" ? (value as Record<string, unknown>) : {};
  const rawDecisions = row.decisions && typeof row.decisions === "object"
    ? (row.decisions as Record<string, unknown>)
    : {};
  const decisions: Record<string, EntityResolutionDecision> = {};
  for (const [rawKey, rawValue] of Object.entries(rawDecisions)) {
    const decision = normalizeDecision(rawValue);
    const key = normalizeEntityResolutionIdentity(rawKey || decision?.canonicalName);
    if (!key || !decision) continue;
    decisions[key] = decision;
  }
  return {
    schemaVersion: 1,
    generatedAt: text(row.generatedAt, 80),
    decisions,
  };
}

export const entityResolutionDecisionManifest = normalizeEntityResolutionManifest(rawDecisions);

function addIndex<T>(index: Map<string, T[]>, alias: string, value: T) {
  const key = normalizeEntityResolutionIdentity(alias);
  if (!key) return;
  const values = index.get(key) ?? [];
  if (!values.includes(value)) index.set(key, [...values, value]);
}

const companyIndex = new Map<string, typeof companyRegistryEntries>();
for (const company of companyRegistryEntries) {
  for (const alias of [company.name, company.englishName ?? "", company.slug, ...company.aliases]) {
    addIndex(companyIndex, alias, company);
  }
}

type ResolutionPerson = {
  slug: string;
  name: string;
  englishName?: string;
  aliases?: string[];
};

function configuredPersonName(raw: string) {
  const value = text(raw, 160);
  const base = text(value.replace(/\s+@[A-Za-z0-9_]+$/u, ""), 160);
  if (!base || base.startsWith("The ")) return "";
  if (/[\u3400-\u9fff]/u.test(base)) {
    return /^[\u3400-\u9fff·•]{2,8}$/u.test(base) ? base : "";
  }
  const parts = base.split(/\s+/u);
  if (parts.length < 2 || parts.length > 5) return "";
  return parts.every((part) => /^[A-Z][a-z]+(?:[-'][A-Z]?[a-z]+)*$/u.test(part))
    ? base
    : "";
}

const personIndex = new Map<string, ResolutionPerson[]>();
for (const person of catalogPeople) {
  for (const alias of [person.name, person.englishName, person.slug]) {
    addIndex(personIndex, alias, person);
  }
}
for (const track of userTrackingConfig.tracks) {
  for (const rawName of track.people) {
    const canonicalName = configuredPersonName(rawName);
    const key = normalizeEntityResolutionIdentity(canonicalName);
    if (!canonicalName || !key || personIndex.has(key)) continue;
    const person: ResolutionPerson = {
      slug: key,
      name: canonicalName,
      englishName: canonicalName,
      aliases: [rawName],
    };
    for (const alias of [canonicalName, rawName]) {
      addIndex(personIndex, alias, person);
    }
  }
}

const topicIndex = new Map<string, string[]>();
for (const track of userTrackingConfig.tracks) {
  for (const alias of [track.name, ...track.keywords]) {
    const value = text(alias, 160);
    if (!value) continue;
    addIndex(topicIndex, value, value);
  }
}

function decisionIndex(manifest: EntityResolutionManifest) {
  const index = new Map<string, EntityResolutionDecision>();
  for (const [key, decision] of Object.entries(manifest.decisions)) {
    index.set(key, decision);
    for (const alias of [decision.canonicalName, ...decision.aliases]) {
      const aliasKey = normalizeEntityResolutionIdentity(alias);
      if (aliasKey) index.set(aliasKey, decision);
    }
  }
  return index;
}

function sourceText(source: EntityResolutionSourceContext | undefined) {
  if (!source) return "";
  return text(
    [
      source.title,
      source.summary,
      source.eventType,
      source.sourceName,
      source.channel,
      source.channelLabel,
      source.url,
    ].join(" "),
    4_000,
  );
}

function contextWindow(context: string, name: string) {
  const normalizedContext = context.toLocaleLowerCase("zh-CN");
  const normalizedName = text(name, 200).toLocaleLowerCase("zh-CN");
  const index = normalizedContext.indexOf(normalizedName);
  if (index < 0) return "";
  return context.slice(Math.max(0, index - 90), Math.min(context.length, index + normalizedName.length + 90));
}

const COMPANY_CUES = /公司|企业|集团|平台|实验室|研究院|基金|资本|创投|startup|company|platform|labs?\b|technolog(?:y|ies)\b|systems?\b|capital\b|ventures?\b|foundation\b|inc\.?\b|corp(?:oration)?\b|ltd\.?\b/iu;
const PERSON_CUES = /创始人|联合创始人|CEO|CTO|CFO|董事|高管|教授|研究员|工程师|作者|大佬|先锋|先生|女士|founder|executive|professor|researcher|engineer|author/iu;
const TOPIC_CUES = /编程语言|语言|框架|协议|标准|技术|算法|模型|芯片架构|数据库|开源项目|软件包|工具链|programming language|framework|protocol|standard|technology|algorithm|model|library|package|toolchain/iu;
const COMPANY_NAME_CUES = /(?:\b(?:Inc|Corp|Corporation|Ltd|LLC|Labs|Technologies|Systems|Capital|Ventures|Foundation)\b|公司|集团|资本|基金|科技)$/iu;

function result(
  input: ResolveTrackingEntityInput,
  patch: Omit<TrackingEntityResolution, "requestedType" | "decisionKey" | "reclassified">,
): TrackingEntityResolution {
  const decisionKey = normalizeEntityResolutionIdentity(input.name);
  return {
    ...patch,
    requestedType: input.requestedType,
    decisionKey,
    reclassified: patch.entityType !== input.requestedType,
  };
}

export function resolveTrackingEntity(input: ResolveTrackingEntityInput): TrackingEntityResolution {
  const requestedType = ENTITY_TYPES.includes(input.requestedType) ? input.requestedType : "company";
  const name = text(input.name, 160);
  const normalizedInput = { ...input, requestedType, name };
  const key = normalizeEntityResolutionIdentity(name);
  if (!key) {
    return result(normalizedInput, {
      status: "rejected",
      entityType: requestedType,
      canonicalName: name,
      targetId: "",
      confidence: "low",
      source: "unresolved",
      reason: "名称为空或无法形成稳定实体键。",
    });
  }

  const manifest = input.manifest ?? entityResolutionDecisionManifest;
  const reviewed = decisionIndex(manifest).get(key);
  if (reviewed) {
    return result(normalizedInput, {
      status: reviewed.status,
      entityType: reviewed.entityType,
      canonicalName: reviewed.canonicalName,
      targetId: reviewed.targetId,
      confidence: reviewed.confidence,
      source: "human-decision",
      reason: reviewed.note || "复用版本化人工解析决定。",
    });
  }

  const companies = companyIndex.get(key) ?? [];
  if (companies.length === 1) {
    const company = companies[0];
    return result(normalizedInput, {
      status: "resolved",
      entityType: "company",
      canonicalName: company.name,
      targetId: `company:${company.slug}`,
      confidence: "verified",
      source: "company-registry",
      reason: "名称唯一命中正式公司注册表。",
    });
  }
  if (companies.length > 1) {
    return result(normalizedInput, {
      status: "review",
      entityType: requestedType,
      canonicalName: name,
      targetId: "",
      confidence: "low",
      source: "company-registry",
      reason: "名称同时命中多个正式公司别名，需要人工消歧。",
    });
  }

  const persons = personIndex.get(key) ?? [];
  if (persons.length === 1) {
    const person = persons[0];
    return result(normalizedInput, {
      status: "resolved",
      entityType: "person",
      canonicalName: person.name,
      targetId: `person:${person.slug}`,
      confidence: "high",
      source: "people-registry",
      reason: "名称唯一命中正式人物目录。",
    });
  }
  if (persons.length > 1) {
    return result(normalizedInput, {
      status: "review",
      entityType: requestedType,
      canonicalName: name,
      targetId: "",
      confidence: "low",
      source: "people-registry",
      reason: "名称同时命中多个人物别名，需要人工消歧。",
    });
  }

  if (requestedType === "company") {
    const matches = institutionTrackingMatches(name);
    if (matches.length) {
      const unique = matches.length === 1;
      return result(normalizedInput, {
        status: unique ? "resolved" : "review", entityType: "company",
        canonicalName: unique ? matches[0].name : name,
        targetId: unique ? matches[0].targetId : "", confidence: unique ? "verified" : "low",
        source: "institution-directory",
        reason: unique ? "名称唯一命中已核验投资机构目录；仅解析机构身份，不代表公司档案发布。" : "名称命中多个投资机构，需要人工消歧。",
      });
    }
  }
  const topics = topicIndex.get(key) ?? [];
  if (topics.length === 1) {
    return result(normalizedInput, {
      status: "resolved",
      entityType: "topic",
      canonicalName: topics[0],
      targetId: `topic:${key}`,
      confidence: "high",
      source: "tracking-taxonomy",
      reason: "名称唯一命中已审核追踪赛道或关键词。",
    });
  }

  if (requestedType === "person" || requestedType === "topic") {
    return result(normalizedInput, {
      status: "resolved",
      entityType: requestedType,
      canonicalName: name,
      targetId: `${requestedType}:${key}`,
      confidence: "medium",
      source: "explicit-type",
      reason: `管理员显式选择“${requestedType === "person" ? "人物" : "技术／主题"}”，且未与正式目录冲突。`,
    });
  }

  const context = sourceText(input.source);
  const local = contextWindow(context, name);
  if (PERSON_CUES.test(local) || TOPIC_CUES.test(local)) {
    return result(normalizedInput, {
      status: "review",
      entityType: requestedType,
      canonicalName: name,
      targetId: "",
      confidence: "low",
      source: "source-context",
      reason: PERSON_CUES.test(local)
        ? "原文邻近语境更像人物，不能直接作为公司写入。"
        : "原文邻近语境更像技术、项目或工具，不能直接作为公司写入。",
    });
  }
  if (COMPANY_CUES.test(local) || COMPANY_NAME_CUES.test(name)) {
    return result(normalizedInput, {
      status: "resolved",
      entityType: "company",
      canonicalName: name,
      targetId: `company-candidate:${key}`,
      confidence: "medium",
      source: "source-context",
      reason: "原文邻近语境存在公司、企业或平台表述，可进入候选审核。",
    });
  }

  return result(normalizedInput, {
    status: "review",
    entityType: requestedType,
    canonicalName: name,
    targetId: "",
    confidence: "low",
    source: "unresolved",
    reason: "缺少正式目录命中或明确公司语境，暂存到实体解析审核队列。",
  });
}

export function normalizeTrackingEntityResolution(value: unknown): TrackingEntityResolution | undefined {
  if (!value || typeof value !== "object") return undefined;
  const row = value as Record<string, unknown>;
  const status = STATUSES.includes(row.status as EntityResolutionStatus)
    ? (row.status as EntityResolutionStatus)
    : null;
  const requestedType = ENTITY_TYPES.includes(row.requestedType as EntityResolutionEntityType)
    ? (row.requestedType as EntityResolutionEntityType)
    : null;
  const entityType = ENTITY_TYPES.includes(row.entityType as EntityResolutionEntityType)
    ? (row.entityType as EntityResolutionEntityType)
    : null;
  const canonicalName = text(row.canonicalName, 160);
  if (!status || !requestedType || !entityType || !canonicalName) return undefined;
  return {
    status,
    requestedType,
    entityType,
    canonicalName,
    targetId: text(row.targetId, 240),
    confidence: CONFIDENCES.includes(row.confidence as EntityResolutionConfidence)
      ? (row.confidence as EntityResolutionConfidence)
      : "low",
    source: [
      "human-decision",
      "company-registry",
      "institution-directory",
      "people-registry",
      "tracking-taxonomy",
      "source-context",
      "explicit-type",
      "unresolved",
    ].includes(String(row.source))
      ? (row.source as EntityResolutionSource)
      : "unresolved",
    reason: text(row.reason, 600),
    decisionKey: normalizeEntityResolutionIdentity(row.decisionKey || canonicalName),
    reclassified: Boolean(row.reclassified ?? entityType !== requestedType),
  };
}

File: tools/entity_resolution.py
#!/usr/bin/env python3
"""Resolve captured names into versioned company, person or topic entities."""

from __future__ import annotations

import json
import re
from dataclasses import dataclass, asdict
from pathlib import Path
from typing import Any, Mapping

ROOT = Path(__file__).resolve().parents[1]
DECISIONS_PATH = ROOT / "config" / "entity_resolution_decisions.json"
COMPANY_REGISTRY_PATH = ROOT / "config" / "company_registry.json"
PEOPLE_PATH = ROOT / "public" / "data" / "people.json"
TRACKING_PATH = ROOT / "config" / "user_tracking.json"

ENTITY_TYPES = {"company", "person", "topic"}
STATUSES = {"resolved", "review", "rejected"}
CONFIDENCES = {"verified", "high", "medium", "low"}

COMPANY_CUES = re.compile(
    r"公司|企业|集团|平台|实验室|研究院|基金|资本|创投|startup|company|platform|labs?\b|"
    r"technolog(?:y|ies)\b|systems?\b|capital\b|ventures?\b|foundation\b|inc\.?\b|"
    r"corp(?:oration)?\b|ltd\.?\b",
    re.IGNORECASE,
)
PERSON_CUES = re.compile(
    r"创始人|联合创始人|CEO|CTO|CFO|董事|高管|教授|研究员|工程师|作者|大佬|先锋|先生|女士|"
    r"founder|executive|professor|researcher|engineer|author",
    re.IGNORECASE,
)
TOPIC_CUES = re.compile(
    r"编程语言|语言|框架|协议|标准|技术|算法|模型|芯片架构|数据库|开源项目|软件包|工具链|"
    r"programming language|framework|protocol|standard|technology|algorithm|model|library|package|toolchain",
    re.IGNORECASE,
)
COMPANY_NAME_CUES = re.compile(
    r"(?:\b(?:Inc|Corp|Corporation|Ltd|LLC|Labs|Technologies|Systems|Capital|Ventures|Foundation)\b|"
    r"公司|集团|资本|基金|科技)$",
    re.IGNORECASE,
)


def clean(value: Any, limit: int = 1200) -> str:
    return re.sub(r"\s+", " ", str(value or "")).strip()[:limit]


def normalize_identity(value: Any) -> str:
    return re.sub(r"[^a-z0-9\u3400-\u9fff]+", "", clean(value, 300).casefold())


def load_json(path: Path, fallback: Any) -> Any:
    try:
        return json.loads(path.read_text(encoding="utf-8"))
    except (OSError, json.JSONDecodeError):
        return fallback


@dataclass(frozen=True)
class Resolution:
    status: str
    requestedType: str
    entityType: str
    canonicalName: str
    targetId: str
    confidence: str
    source: str
    reason: str
    decisionKey: str
    reclassified: bool

    def to_dict(self) -> dict[str, Any]:
        return asdict(self)


def _result(
    *,
    requested_type: str,
    name: str,
    status: str,
    entity_type: str,
    canonical_name: str,
    target_id: str = "",
    confidence: str = "low",
    source: str = "unresolved",
    reason: str = "",
) -> Resolution:
    return Resolution(
        status=status,
        requestedType=requested_type,
        entityType=entity_type,
        canonicalName=canonical_name,
        targetId=target_id,
        confidence=confidence,
        source=source,
        reason=reason,
        decisionKey=normalize_identity(name),
        reclassified=entity_type != requested_type,
    )


def normalize_decision_manifest(payload: Any) -> dict[str, dict[str, Any]]:
    raw = payload.get("decisions", {}) if isinstance(payload, dict) else {}
    if not isinstance(raw, dict):
        return {}
    decisions: dict[str, dict[str, Any]] = {}
    for raw_key, value in raw.items():
        if not isinstance(value, dict):
            continue
        status = clean(value.get("status"), 20)
        requested_type = clean(value.get("requestedType"), 20) or "company"
        entity_type = clean(value.get("entityType"), 20) or requested_type
        canonical_name = clean(value.get("canonicalName"), 160)
        if (
            status not in STATUSES
            or requested_type not in ENTITY_TYPES
            or entity_type not in ENTITY_TYPES
            or not canonical_name
        ):
            continue
        row = {
            "status": status,
            "requestedType": requested_type,
            "entityType": entity_type,
            "canonicalName": canonical_name,
            "targetId": clean(value.get("targetId"), 240),
            "aliases": [clean(alias, 240) for alias in value.get("aliases", []) if clean(alias, 240)]
            if isinstance(value.get("aliases"), list)
            else [],
            "confidence": clean(value.get("confidence"), 20)
            if clean(value.get("confidence"), 20) in CONFIDENCES
            else ("verified" if status == "resolved" else "low"),
            "note": clean(value.get("note"), 600),
        }
        keys = {normalize_identity(raw_key), normalize_identity(canonical_name)}
        keys.update(normalize_identity(alias) for alias in row["aliases"])
        for key in keys:
            if key:
                decisions[key] = row
    return decisions


def _add_index(index: dict[str, list[dict[str, Any]]], alias: Any, value: dict[str, Any]) -> None:
    key = normalize_identity(alias)
    if not key:
        return
    rows = index.setdefault(key, [])
    if value not in rows:
        rows.append(value)


def company_index(payload: Any) -> dict[str, list[dict[str, Any]]]:
    companies = payload.get("companies", []) if isinstance(payload, dict) else []
    index: dict[str, list[dict[str, Any]]] = {}
    for company in companies if isinstance(companies, list) else []:
        if not isinstance(company, dict):
            continue
        for alias in [
            company.get("name"),
            company.get("englishName"),
            company.get("slug"),
            *(company.get("aliases", []) if isinstance(company.get("aliases"), list) else []),
        ]:
            _add_index(index, alias, company)
    return index


def people_index(payload: Any) -> dict[str, list[dict[str, Any]]]:
    people = payload.get("people", []) if isinstance(payload, dict) else []
    index: dict[str, list[dict[str, Any]]] = {}
    for person in people if isinstance(people, list) else []:
        if not isinstance(person, dict):
            continue
        for alias in [
            person.get("name"),
            person.get("englishName"),
            person.get("slug"),
            *(person.get("aliases", []) if isinstance(person.get("aliases"), list) else []),
        ]:
            _add_index(index, alias, person)
    return index


def topic_index(payload: Any) -> dict[str, list[str]]:
    tracks = payload.get("tracks", []) if isinstance(payload, dict) else []
    index: dict[str, list[str]] = {}
    for track in tracks if isinstance(tracks, list) else []:
        if not isinstance(track, dict):
            continue
        values = [track.get("name")]
        if isinstance(track.get("keywords"), list):
            values.extend(track["keywords"])
        for raw in values:
            value = clean(raw, 160)
            key = normalize_identity(value)
            if not key or not value:
                continue
            rows = index.setdefault(key, [])
            if value not in rows:
                rows.append(value)
    return index


def source_text(source: Mapping[str, Any] | None) -> str:
    if not source:
        return ""
    return clean(
        " ".join(
            clean(source.get(key), 1200)
            for key in ("title", "summary", "eventType", "sourceName", "channel", "channelLabel", "url")
        ),
        4000,
    )


def context_window(context: str, name: str) -> str:
    index = context.casefold().find(clean(name, 200).casefold())
    if index < 0:
        return ""
    return context[max(0, index - 90) : min(len(context), index + len(name) + 90)]


def resolve_entity(
    requested_type: str,
    name: str,
    source: Mapping[str, Any] | None = None,
    *,
    decisions_payload: Any | None = None,
    company_registry_payload: Any | None = None,
    people_payload: Any | None = None,
    tracking_payload: Any | None = None,
) -> Resolution:
    requested_type = requested_type if requested_type in ENTITY_TYPES else "company"
    name = clean(name, 160)
    key = normalize_identity(name)
    if not key:
        return _result(
            requested_type=requested_type,
            name=name,
            status="rejected",
            entity_type=requested_type,
            canonical_name=name,
            reason="名称为空或无法形成稳定实体键。",
        )

    decisions = normalize_decision_manifest(
        decisions_payload if decisions_payload is not None else load_json(DECISIONS_PATH, {})
    )
    reviewed = decisions.get(key)
    if reviewed:
        return _result(
            requested_type=requested_type,
            name=name,
            status=reviewed["status"],
            entity_type=reviewed["entityType"],
            canonical_name=reviewed["canonicalName"],
            target_id=reviewed["targetId"],
            confidence=reviewed["confidence"],
            source="human-decision",
            reason=reviewed["note"] or "复用版本化人工解析决定。",
        )

    companies = company_index(
        company_registry_payload
        if company_registry_payload is not None
        else load_json(COMPANY_REGISTRY_PATH, {})
    ).get(key, [])
    if len(companies) == 1:
        company = companies[0]
        return _result(
            requested_type=requested_type,
            name=name,
            status="resolved",
            entity_type="company",
            canonical_name=clean(company.get("name"), 160),
            target_id=f"company:{clean(company.get('slug'), 160)}",
            confidence="verified",
            source="company-registry",
            reason="名称唯一命中正式公司注册表。",
        )
    if len(companies) > 1:
        return _result(
            requested_type=requested_type,
            name=name,
            status="review",
            entity_type=requested_type,
            canonical_name=name,
            source="company-registry",
            reason="名称同时命中多个正式公司别名，需要人工消歧。",
        )

    persons = people_index(
        people_payload if people_payload is not None else load_json(PEOPLE_PATH, {})
    ).get(key, [])
    if len(persons) == 1:
        person = persons[0]
        return _result(
            requested_type=requested_type,
            name=name,
            status="resolved",
            entity_type="person",
            canonical_name=clean(person.get("name"), 160),
            target_id=f"person:{clean(person.get('slug') or person.get('id'), 160)}",
            confidence="high",
            source="people-registry",
            reason="名称唯一命中正式人物目录。",
        )
    if len(persons) > 1:
        return _result(
            requested_type=requested_type,
            name=name,
            status="review",
            entity_type=requested_type,
            canonical_name=name,
            source="people-registry",
            reason="名称同时命中多个人物别名，需要人工消歧。",
        )

    # Institution identity is distinct from operating-company publication.
    if requested_type == "company":
        try:
            from .institution_identity import lookup as institution_lookup
        except ImportError:
            from institution_identity import lookup as institution_lookup
        institutions = institution_lookup(name)
        if institutions:
            unique = len(institutions) == 1
            institution = institutions[0]
            return _result(
                requested_type=requested_type, name=name,
                status="resolved" if unique else "review", entity_type="company",
                canonical_name=institution["name"] if unique else name,
                target_id=institution["targetId"] if unique else "",
                confidence="verified" if unique else "low", source="institution-directory",
                reason=("名称唯一命中已核验投资机构目录；仅解析机构身份，不代表公司档案发布。"
                        if unique else "名称命中多个投资机构，需要人工消歧。"),
            )

    topics = topic_index(
        tracking_payload if tracking_payload is not None else load_json(TRACKING_PATH, {})
    ).get(key, [])
    if len(topics) == 1:
        return _result(
            requested_type=requested_type,
            name=name,
            status="resolved",
            entity_type="topic",
            canonical_name=topics[0],
            target_id=f"topic:{key}",
            confidence="high",
            source="tracking-taxonomy",
            reason="名称唯一命中已审核追踪赛道或关键词。",
        )

    if requested_type in {"person", "topic"}:
        label = "人物" if requested_type == "person" else "技术／主题"
        return _result(
            requested_type=requested_type,
            name=name,
            status="resolved",
            entity_type=requested_type,
            canonical_name=name,
            target_id=f"{requested_type}:{key}",
            confidence="medium",
            source="explicit-type",
            reason=f"管理员显式选择“{label}”，且未与正式目录冲突。",
        )

    context = source_text(source)
    local = context_window(context, name)
    if PERSON_CUES.search(local) or TOPIC_CUES.search(local):
        return _result(
            requested_type=requested_type,
            name=name,
            status="review",
            entity_type=requested_type,
            canonical_name=name,
            confidence="low",
            source="source-context",
            reason=(
                "原文邻近语境更像人物，不能直接作为公司写入。"
                if PERSON_CUES.search(local)
                else "原文邻近语境更像技术、项目或工具，不能直接作为公司写入。"
            ),
        )
    if COMPANY_CUES.search(local) or COMPANY_NAME_CUES.search(name):
        return _result(
            requested_type=requested_type,
            name=name,
            status="resolved",
            entity_type="company",
            canonical_name=name,
            target_id=f"company-candidate:{key}",
            confidence="medium",
            source="source-context",
            reason="原文邻近语境存在公司、企业或平台表述，可进入候选审核。",
        )

    return _result(
        requested_type=requested_type,
        name=name,
        status="review",
        entity_type=requested_type,
        canonical_name=name,
        confidence="low",
        source="unresolved",
        reason="缺少正式目录命中或明确公司语境，暂存到实体解析审核队列。",
    )
